### PR TITLE
Fix EventPipe.Enable() to be aware of multiple EventPipe sessions

### LINKED ISSUE #34839: Enabling EventPipe via the managed EventPipe class isn't thread-safe
Reported by @xiaomi7732:

```cs
using System;
namespace System.Diagnostics.Tracing
{
    internal static class EventPipe
    {
        private static ulong s_sessionID;
        internal static void Enable(EventPipeConfiguration configuration)
        {
            if (configuration == null)
            {
                throw new System.ArgumentNullException("configuration");
            }
            if (configuration.Providers == null)
            {
                throw new System.ArgumentNullException("Providers");
            }
            EventPipeProviderConfiguration[] providers = configuration.Providers;
            EventPipe.s_sessionID = EventPipeInternal.Enable(configuration.OutputFile, configuration.Format, configuration.CircularBufferSizeInMB, providers, (uint)providers.Length);
        }
        internal static void Disable()
        {
            EventPipeInternal.Disable(EventPipe.s_sessionID);
        }
    }
}
```

This code may leak an EventPipe session when EventPipe.Enable() is called concurrently where two session may be enabled from different codepaths and both of them will be enabled, but only one of them will be tracked via s_sessionID meaning whichever session isn't tracked will be leaked when Disable() is called.

On top of that, even without involving concurrency, it's also possible for us to never call Disable() on a session that was overwritten by calling Enable() twice before the first one was disabled.

### PR DESCRIPTION
**Pull request by sywhang**
This fixes #34839

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipe.cs [hunk 1]
```diff
@@ -177,7 +178,7 @@ internal void SetProfilerSamplingRate(TimeSpan minTimeBetweenSamples)
 
     internal static class EventPipe
     {
-        private static ulong s_sessionID = 0;
+        internal static ConcurrentQueue<ulong> s_sessionIDs = new ConcurrentQueue<ulong>();
```
Comment (josalem) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipe.cs:181]: Could we use a `HashSet` or just a `List` and expose the ability to disable a specific session or at least the number of active sessions?  This change currently expects users of this API to keep track of how many sessions they open, and more importantly, expects there to only be _one_ user.  If there are multiple callers to this API and they interleave calls to `Enable` and `Disable`, they could end up disabling each other's sessions.

e.g.,
1. Caller1 calls `Enable`
2. Caller2 calls `Enable`
3. Caller2 calls `Disable` and they just disabled Caller1's session

**Review by josalem — DISMISSED** [review]
~~I think we should add some sort of mechanism to prevent multiple callers from disabling each other's sessions (see comment).  We don't _expect_ there to be external users, but nothing stops someone from getting at this API via reflection.~~  Chatted offline, this level of control is unnecessary here :smile:

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipe.cs (hunk 1 above)
Comment (sywhang) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipe.cs:181]: Yup. I thought about this too, and discussed options available with @xiaomi7732 from AppInsights Service Profiler team. The problem with doing that is that it changes the method signature, which breaks back-compat with 2.2 which is still supported by AppInsights ServiceProfiler (which, afaik, is the only team that enables EventPipe using private reflection). 

It's not ideal and there is the case you mentioned which would ideally break. But customers shouldn't be (and were never encouraged by us via any blog articles, docs, etc.) to enable/disable EventPipe this way, so I'm hoping that people won't run into such issues, and when they do, changing their code to not use reflection will be actually easier than whatever they were doing before :)

File: src/coreclr/tests/src/tracing/regress/GitHub_34839/GitHub_34839.cs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            }
+
+            // Enable tracing many times
+            for (int i = 0; i < configurationCnt; i++)
+            {
+                Enable(configs[i]);
+            }
+
+            // Sleep for 5 second to let some events to be written 
+            Thread.Sleep(1000);
```
Comment (josalem) [inline src/coreclr/tests/src/tracing/regress/GitHub_34839/GitHub_34839.cs:216]: Comment doesn't match code.  Also, does the sleep need to be 1s or could it be something shorter, like 500ms?

**Review by josalem — APPROVED** [review]
small nit on the test, but otherwise looks good.

**Review by noahfalk — CHANGES_REQUESTED** [review]
#34839 suggests that this API was intended to be thread-safe and capable of supporting multiple sessions. I doubt either of those traits were in the original design criteria and it makes me question if the design used here where we retroactively change the API semantics is the best overall choice for how to solve the problem. I'm not sure if there was already discussion about how to solve this but I couldn't find anything written in the issue or the PR here that answered some of the questions that came to mind:
- What value does Service Profiler get from multiple concurrent sessions?
- Does the API need to be safe for free-threaded access or can Service Profiler do their own synchronization?
- Is there any existing functionality that could be used to achieve Service Profiler's goal (such as NetCore.Client APIs)? If the existing tech doesn't work well for the scenario, why not?
- Any other design options we've looked at and what was appealing about doing it this way vs. alternatives?

**Comment by stephentoub** [comment]
@sywhang, are you still working on this?  Thanks.

**Pull request closed without merging**